File: Matlab/Heston/TreatReturns.m
function Res = TreatReturns(Model,M,n,T,kappa,theta,sigma,mu,rho,v0,momth,setup)

    if strcmp(setup, 'C')

        if     strcmp(Model, 'EulerFT')
            
            tic
            [V, Z] = fEulTruncV(M,n,T,kappa,theta,sigma,v0);
            X = fEulTruncRet(M,n,T,V,Z,mu,rho);
            R = X(end,:) - X(1,:);
            etime = toc;
            
        elseif strcmp(Model, 'Zhu') 

            tic
            [V, Z] = fZhuV(M,n,T,kappa,theta,sigma,v0);
            X = fZhuRet(M,n,T,V,Z,mu,rho);
            R = X(end,:) - X(1,:);
            etime = toc;

        elseif strcmp(Model, 'KJ') 

            tic
            [V, Z] = fKJV(M,n,T,kappa,theta,sigma,v0);
            X = fKJRet(M,n,T,V,Z,sigma,mu,rho);
            R = X(end,:) - X(1,:);
            etime = toc;

        elseif strcmp(Model, 'QE') 

            tic
            V = fQEV(M,n,T,kappa,theta,sigma,v0);
            X = fQERet(M,n,T,V,kappa,theta,sigma,mu,rho);  
            R = X(end,:) - X(1,:);
            etime = toc;

        elseif strcmp(Model, 'BK') 

            tic
            V = fSUBV(M,n,T,kappa,theta,sigma, v0);
            X = fBKRet(M,n,T,V,kappa,theta,sigma,mu,rho);
            R = X(end,:) - X(1,:);
            etime = toc;

        elseif strcmp(Model, 'Smith') && T / n >= 1 / 12 
            
            tic
            V = fSUBV(M,n,T,kappa,theta,sigma, v0);
            [zgrid, m1Sv, stdISv, hzv, Nzv, fv] = simSmithInit(kappa, theta, sigma, T, n);
            X = fSmithRet(M,n,T,V,kappa,theta,sigma,mu,rho, zgrid, hzv, Nzv, fv, m1Sv, stdISv);
            R = X(end,:) - X(1,:);
            etime = toc;

        elseif strcmp(Model, 'GKPap') 
     
            tic
            V = fSUBV(M,n,T,kappa,theta,sigma, v0);
            [nu, Cz, K, lambda_n, gamma_n, i_X2, i_Z, q_i_X2, q_i_Z, mcp, I_X2, I_Z] = simGKInit(T,n,kappa,theta,sigma);
            X = fGKPap2(M, n, T, V, kappa, theta, sigma, mu, rho, nu, Cz, K, lambda_n, gamma_n, i_X2, i_Z, q_i_X2, q_i_Z, mcp, I_X2, I_Z);
            R = X(end,:) - X(1,:);
            etime = toc;

        elseif strcmp(Model, 'TW') 
            
            tic
            V = fSUBV(M,n,T,kappa,theta,sigma, v0); % Use exact scheme instead of IPZ scheme to draw variance, since IPZ is unreasonable slow
            [C1, C2, grid, EX2,sig2X2, EIV_star, VarIV_star] = simTWInit(n,T,kappa,theta,sigma);
            X = fTWRet(M, n, T, V, kappa, theta, sigma, mu, rho, grid, C1, C2, EIV_star, VarIV_star, EX2, sig2X2);
            R = X(end,:) - X(1,:);
            etime = toc;

        elseif strcmp(Model, 'BBG')  

            tic
            V = fSUBV(M,n,T,kappa,theta,sigma, v0);
            % Choose zeta "wisely". If e. g. M is small (e.g. 1e3), zeta =
            % 85 will result in a non available grid -> choose zeta  = 80
            if 2 * kappa * theta / sigma^2 < 1 
                zeta = 60;
            else
                zeta = 80;
            end
            [C1, C2, grid, EIV_star, VarIV_star] = simBBGInit(M, n, T, kappa, theta, sigma, zeta);
            X = fBBGRet(M,n,T,V,kappa,theta,sigma,mu,rho, C1, C2, grid, EIV_star, VarIV_star);
            R = X(end,:) - X(1,:);
            etime = toc;

        elseif strcmp(Model, 'ORS')
            
            mode = 'unconditional'; % Here, write unconditional since even for the UC setup, we use the 'C' setup mentioned here...
            target = 'skew';

            tic
            V = fQEV(M,n,T,kappa,theta,sigma,v0);
            X = fORSRet(M,n,T,V,kappa,theta,sigma,mu,rho,target,mode);
            R = X(end,:) - X(1,:);
            etime = toc;
            
        elseif strcmp(Model, 'ORS + SUB')

            mode = 'unconditional';  % Here, write unconditional since even for the UC setup, we use the 'C' setup mentioned here...
            target = 'skew';

            tic
            V = fSUBV(M,n,T,kappa,theta,sigma,v0);
            X = fORSRet(M,n,T,V,kappa,theta,sigma,mu,rho,target,mode);
            R = X(end,:) - X(1,:);
            etime = toc;
            
        end   

    elseif strcmp(setup, 'UC')
        
        if     strcmp(Model, 'EulerFT')
            
            tic
            [V, Z] = fEulTruncV(M,n,T,kappa,theta,sigma,v0);
            X = fEulTruncRet(M,n,T,V,Z,sigma,mu,rho);
            R = X(: ,2:end) - X(: , 1:end-1);
            etime = toc;
            
        elseif strcmp(Model, 'Zhu') 

            tic
            [V, Z] = fZhuV(M,n,T,kappa,theta,sigma,v0);
            X = fZhuRet(M,n,T,V,Z,sigma,mu,rho);
            R = X(: ,2:end) - X(: , 1:end-1);
            etime = toc;

        elseif strcmp(Model, 'KJ') 

            tic
            [V, Z] = fKJV(M,n,T,kappa,theta,sigma,v0);
            X = fKJRet(M,n,T,V,Z,sigma,mu,rho);
            R = X(: ,2:end) - X(: , 1:end-1);
            etime = toc;

        elseif strcmp(Model, 'QE') 

            tic
            V = fQEV(M,n,T,kappa,theta,sigma,v0);
            X = fQERet(M,n,T,V,kappa,theta,sigma,mu,rho);  
            R = X(: ,2:end) - X(: , 1:end-1);
            etime = toc;

        elseif strcmp(Model, 'BK') 

            tic
            V = fSUBV(M,n,T,kappa,theta,sigma, v0);
            X = fBKRet(M,n,T,V,kappa,theta,sigma,mu,rho);
            R = X(: ,2:end) - X(: , 1:end-1);
            etime = toc;

        elseif strcmp(Model, 'Smith') && T/n >= 1 / 12 
            
            tic
            V = fSUBV(M,n,T,kappa,theta,sigma, v0);
            [zgrid, m1Sv, stdISv, hzv, Nzv, fv] = simSmithInit(kappa, theta, sigma, T, n);
            X = fSmithRet(M,n,T,V,kappa,theta,sigma,mu,rho, zgrid, hzv, Nzv, fv, m1Sv, stdISv);
            R = X(: ,2:end) - X(: , 1:end-1);
            etime = toc;

        elseif strcmp(Model, 'GKPap') 

            tic
            V = fSUBV(M,n,T,kappa,theta,sigma, v0);
            [nu, Cz, K, lambda_n, gamma_n, i_X2, i_Z, q_i_X2, q_i_Z, mcp, I_X2, I_Z] = simGKInit(T,n,kappa,theta,sigma);
            X = fGKPap2(M, n, T, V, kappa, theta, sigma, mu, rho, nu, Cz, K, lambda_n, gamma_n, i_X2, i_Z, q_i_X2, q_i_Z, mcp, I_X2, I_Z);
            R = X(: ,2:end) - X(: , 1:end-1);
            etime = toc;

        elseif strcmp(Model, 'TW') 

            tic
            V = fSUBV(M,n,T,kappa,theta,sigma, v0); % Use exact scheme instead of IPZ scheme to draw variance, since IPZ is unreasonable slow
            [C1, C2, grid, EX2,sig2X2, EIV_star, VarIV_star] = simTWInit(n,T,kappa,theta,sigma);
            X = fTWRet(M, n, T, V, kappa, theta, sigma, mu, rho, grid, C1, C2, EIV_star, VarIV_star, EX2, sig2X2);
            R = X(: ,2:end) - X(: , 1:end-1);
            etime = toc;
            

        elseif strcmp(Model, 'BBG')  

            tic
            V = fSUBV(M,n,T,kappa,theta,sigma, v0);
            [C1, C2, grid, EIV_star, VarIV_star] = simBBGInit(M, n, T, kappa, theta, sigma, 80);
            X = fBBGRet(M,n,T,V,kappa,theta,sigma,mu,rho, C1, C2, grid, EIV_star, VarIV_star);
            R = X(: ,2:end) - X(: , 1:end-1);
            etime = toc;

        elseif strcmp(Model, 'ORS')
            
            mode = 'unconditional';
            target = 'skew';

            tic
            V = fQEV(M,n,T,kappa,theta,sigma,v0);
            X = fORSRet(M,n,T,V,kappa,theta,sigma,mu,rho,v0,target,mode);
            R = X(: ,2:end) - X(: , 1:end-1);
            etime = toc;
            
        elseif strcmp(Model, 'ORS + SUB')

            mode = 'unconditional';
            target = 'skew';

            tic
            V = fSUBV(M,n,T,kappa,theta,sigma,v0);
            X = fORSRet(M,n,T,V,kappa,theta,sigma,mu,rho,v0,target,mode);
            R = X(: ,2:end) - X(: , 1:end-1);
            etime = toc;
            
        end   
         
    end
    
   Errors = getStats(R, momth);
   Res = [Errors etime];
   
end

File: R/Heston/TreatReturns.r
TreatReturns = function(Model, S0, v0, kappa, theta, sigma, mu, rho, T, n, M, momth, setup){
  
  if(setup == 'C'){
    
    if(Model == 'EulerFT'){ 
      
      tic = Sys.time()
      rT = EulerScheme(S0, v0, kappa, theta, sigma, mu, rho, T, n, M, mode = "FT")[,n+1] - log(S0)
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
    }else if(Model == 'KJ'){ 
      
      tic = Sys.time()
      rT = IKJIMMScheme(S0, v0, kappa, theta, sigma, mu, rho, T, n, M)[,n+1] - log(S0)
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
     }else if(Model == 'QE'){  
      
      tic = Sys.time()
      rT = QEScheme(S0, v0, kappa, theta, sigma, mu, rho, T, n, M, gamma1  = 0.5, gamma2 = 0.5)[,n+1] - log(S0)
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
    }else if(Model == 'BK'){ 
      
      tic = Sys.time()
      rT = BKScheme(S0, v0, kappa, theta, sigma, mu, rho, T, n, M)[,n+1] - log(S0)
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
    }else if(Model == 'Smith' && T / n >= 1 / 12){ 
    
      tic = Sys.time()
      rT = Smith(S0, v0, kappa, theta, sigma, mu, rho, T, n, M)[,n+1] - log(S0)
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
    }else if(Model == 'GK'){ 
      
      tic = Sys.time()
      rT = GEScheme(S0, v0, kappa, theta, sigma, mu, rho, T, n, M)[,n+1] - log(S0)
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
    }else if(Model == 'TW'){ 
      
      tic = Sys.time()
      v  = Exact(v0, kappa, theta, sigma, T, n, M) #Use the exact method instead of the IPZ - scheme since this approach will take forever
      rT = IGScheme(S0, v0, kappa, theta, sigma, mu, rho, T, n, M, v = v)[,n+1] - log(S0)
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
    }else if(Model == 'BBG'){ 
      
      tic = Sys.time()
      rT = GAScheme(S0, v0, kappa, theta, sigma, mu, rho, T, n, M, zeta = 80)[,n+1] - log(S0)
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
    }else if(Model == 'ORS'){ 
      
      tic = Sys.time()
      rT = QEScheme(S0, v0, kappa, theta, sigma, mu, rho, T, n, M, optimized = TRUE, target = 'skew', mode = 'unconditional')[,n+1] - log(S0)  
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
    }else if(Model == 'ORS + SUB'){ 
      
      tic = Sys.time()
      v  = Exact(v0, kappa, theta, sigma, T, n, M)
      rT = QEScheme(S0, v0, kappa, theta, sigma, mu, rho, T, n, M, optimized = TRUE, target = 'skew', mode = 'conditional', v = v)[,n+1] - log(S0)  
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
    }else if(Model == 'Zhu'){
      
      tic = Sys.time()
      rT = ZhuScheme(S0, v0, kappa, theta, sigma, mu, rho, T, n, M)[,n+1] - log(S0)
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
    }else{
      
      rT = rep(-99, M)
      
    }
  
  }else if(setup == 'UC'){
    
    if(Model == 'EulerFT'){ 
      
      tic = Sys.time()
      rT = diff(EulerScheme(S0, v0, kappa, theta, sigma, mu, rho, T, n, M, mode = "FT")[1,])
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
    }else if(Model == 'KJ'){ 
      
      tic = Sys.time()
      rT = diff(IKJIMMScheme(S0, v0, kappa, theta, sigma, mu, rho, T, n, M)[1,])
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
    }else if(Model == 'QE'){  
      
      tic = Sys.time()
      rT = diff(QEScheme(S0, v0, kappa, theta, sigma, mu, rho, T, n, M, gamma1  = 0.5, gamma2 = 0.5)[1,])
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
    }else if(Model == 'BK'){ 
      
      tic = Sys.time()
      rT = diff(BKScheme(S0, v0, kappa, theta, sigma, mu, rho, T, n, M)[1,])
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
    }else if(Model == 'Smith' && T / n >= 1 / 12){ 
      
      tic = Sys.time()
      rT = diff(Smith(S0, v0, kappa, theta, sigma, mu, rho, T, n, M)[1,])
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
    }else if(Model == 'GK'){ 
      
      tic = Sys.time()
      rT = diff(GEScheme(S0, v0, kappa, theta, sigma, mu, rho, T, n, M)[1,])
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
    }else if(Model == 'TW'){ 
      
      tic = Sys.time()
      v  = Exact(v0, kappa, theta, sigma, T, n, M) #Use the exact method instead of the IPZ - scheme since this approach will take forever
      rT = diff(IGScheme(S0, v0, kappa, theta, sigma, mu, rho, T, n, M, v = v)[1,])
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
    }else if(Model == 'BBG'){ 
      
      tic = Sys.time()
      rT = diff(GAScheme(S0, v0, kappa, theta, sigma, mu, rho, T, n, M, zeta = 80)[1,])
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
    }else if(Model == 'ORS'){ 
      
      tic = Sys.time()
      rT = diff(QEScheme(S0, v0, kappa, theta, sigma, mu, rho, T, n, M, optimized = TRUE, target = 'skew', mode = 'conditional')[1,])  
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
    }else if(Model == 'ORS + SUB'){ 
      
      tic = Sys.time()
      v  = Exact(v0, kappa, theta, sigma, T, n, M)
      rT = diff(QEScheme(S0, v0, kappa, theta, sigma, mu, rho, T, n, M, optimized = TRUE, target = 'skew', mode = 'conditional', v = v)[1,]) 
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
    }else if(Model == 'Zhu'){
      
      tic = Sys.time()
      rT = diff(ZhuScheme(S0, v0, kappa, theta, sigma, mu, rho, T, n, M)[1,])
      toc = Sys.time()
      etime = as.numeric(difftime(toc, tic, units = "secs"))
      
    }else{
      
      rT = rep(-99, T)
      
    }
   
  
 }
  
  Bm  = mean(rT) - momth[1]
  Bv  = var(rT) - momth[2]
  Bsk = EnvStats::skewness(rT) - momth[3]
  Bku = EnvStats::kurtosis(rT, excess = F) - momth[4]
  
  APEm  = abs( Bm / momth[1] )
  APEv  = abs( Bv / momth[2] )
  APEsk = abs( Bsk / momth[3] )
  APEku = abs( Bku / momth[4] )
  
  return(c(Bm, Bv, Bsk, Bku, APEm, APEv, APEsk, APEku, etime))


}                  
